### PR TITLE
config: rework validator configs

### PR DESCRIPTION
**Pull request by bmwill**
This patch does a large rework of our validator configs. Specifically it
creates a single `ValidatorConfig` which represents all of the
configuration a Validator needs to start up. It also creates a
light-weight `ValidatorInfo` type which contains information we expect
to know about other validators (stake, public key, network address) and
will eventually be read from on-chain. This is a replacement for the
`AuthorityInfo` type which was a type that all validators have for all
other validators but included some odd information like the on-disk path
to a validator's db which other validators don't need to know about.

This patch also does an initial rework of creating a `NetworkConfig`
which is used when you want to create N validators to run locally.
There's still some additional work that will need to be done to make
this nicer for testing purposes but this is a good start.

### CODE REVIEW

**Comment by bmwill** [comment]
Note I was able to convert everything in a fairly straightforward manner except some of the benchmarking binaries for remote benchmarking. I think those will require some help from @oxade to get done properly. I still plan on doing a bit more in this area immediately once this PR lands but would like to get this landed asap given the vast blast radius it already has.

**Comment by bmwill** [comment]
One thing of note is that now instead of needing to curry around this extra `AuthorityKeys` type you instead just have all the information you need in a `NetworkConfig`, if you fully own the network, and a Validator only has the information that it needs to know about.

File: sui/src/bin/bench_configure.rs
Comment (mystenmark) [inline sui/src/bin/bench_configure.rs:102]: commented code.

Comment (bmwill) [reply sui/src/bin/bench_configure.rs]: Thanks, i've removed it

**Comment by mystenmark** [comment]
Overall this seems like a good idea. Some comments:

- It appears you've switched to using public keys rather than addresses as the identifier of a validator in the config. I think we should probably stick with addresses.
- This is going to require significant changes for deployment to devnet. One issue is that we will need to have a way to generate a network.conf for a solo validator. Unless those configs are now created by hand (I don't know what changes were necessary for when we removed the private keys from the configs).

**Comment by bmwill** [comment]
> * It appears you've switched to using public keys rather than addresses as the identifier of a validator in the config. I think we should probably stick with addresses.

A SuiAddress is derived from a PublicKey so this actually enforces that we don't have an address that mis-matches the public key. Validators are still identified the same.

> * This is going to require significant changes for deployment to devnet. One issue is that we will need to have a way to generate a network.conf for a solo validator. Unless those configs are now created by hand (I don't know what changes were necessary for when we removed the private keys from the configs).

Yes this will and will need to be handled in a separate change. We don't need nor want a network.conf for a solo validator. Instead the validator binary should only accept a ValidatorConfig as a network config contains information about all validators for testing purposes and individual validators shouldn't know all of the config information for other validators.

File: crates/sui-config/src/lib.rs
Comment (longbowlu) [inline crates/sui-config/src/lib.rs:350]: `committee_size` is probably better - `quorum_size` sounds like the `k` in `k out of n`

File: sui/src/sui_commands.rs
Comment (longbowlu) [inline sui/src/sui_commands.rs:331]: repeating what brandon & I discussed offline:
> we’d like to keep the ability to genesis with a pre-specified validator sets (aka with their public keys). This is how it will work for testnet/mainnet and also help with debugability by keeping the same set of validator keys. People will run sui genesis with a genesis config and set up all the dbs etc etc. and we don’t need to socialize two files/config genesis.conf and network.conf

File: crates/sui-config/src/lib.rs
Comment (longbowlu) [inline crates/sui-config/src/lib.rs:32]: remove?

File: crates/sui-config/src/lib.rs
Comment (longbowlu) [inline crates/sui-config/src/lib.rs:77]: curious why `DebugIgnore`?

File: crates/sui-config/src/lib.rs
Comment (longbowlu) [inline crates/sui-config/src/lib.rs:153]: Today's workflow:
1. we collect validator information in `genesis.conf`, socialize it with all genesis validators (there's a keypair which is specific to the concerned validator)
2. each validator runs `sui genesis` and get a `network.conf`. The `network.conf` only has public keys of all validators (plus a keypair of the concerned validator)

**Review by longbowlu — COMMENTED** [review]
thanks for taking time cleaning this up!
This is a relevant issue being tracked #1837 

Main feedback from the 1st round of looking: how do we expect the genesis process looks like for testnet/mainnet? My gut feeling is we will need to share a file that contains all validator's information, and only public information. (some context: Initially the `network.conf` has all the private keys but we moved away to only contain one key pair recently)

File: sui/src/bin/validator.rs
Comment (bmwill) [reply sui/src/bin/validator.rs:66]: I was trying to not adjust the cmdline flags too much so i can leave this to be cleaned up in a future pr.

File: sui/src/bin/validator.rs
Comment (bmwill) [reply sui/src/bin/validator.rs:59]: This is specifically for the case where a config isn't provided and we want to spin up a lonely validator. I'm not so sure i agree with the overall flow in this case but didn't want to change even more things in this PR since its already long

File: crates/sui-config/src/lib.rs
Comment (bmwill) [reply crates/sui-config/src/lib.rs:77]: ConsensusParameters doesn't impl Debug, I plan to fix it but need to do so in the narwhal repo

File: crates/sui-config/src/lib.rs
Comment (bmwill) [reply crates/sui-config/src/lib.rs:153]: The new workflow should be more straight forward:
1. We perform genesis and generate a NetworkConfig which includes N Validator configs
2. We give the corresponding validator config to each validator

File: sui/src/sui_commands.rs
Comment (bmwill) [reply sui/src/sui_commands.rs:331]: in the end we should only need to provide validators with their own validator config.

If we want to fix the names of the validators for debug-ability there is another generate function which takes an RNG so we can just seed it

**Comment by bmwill** [comment]
> Main feedback from the 1st round of looking: how do we expect the genesis process looks like for testnet/mainnet? My gut feeling is we will need to share a file that contains all validator's information, and only public information. (some context: Initially the `network.conf` had all the private keys and we moved away to only contain one key pair recently)

There's still more work to do to get it to where we'd need for testnet/mainet. But we can't have everyone running genesis themselves. At the end of the day genesis is the initial set of objects/checkpoint that will need to be signed by everyone's private keys so we'll need to perform a ceremony for that. I heavily worked on similar logic for diem so i'm working towards a similar process here. Its going to take a few more PRs to get there though ;)

**Review by longbowlu — APPROVED** [review]
@bmwill  can you provide a concise description of how you envision the genesis would happen?

approving to avoid rebase nightmare

**Comment by bmwill** [comment]
@longbowlu I provided a short description in the issue you linked: https://github.com/MystenLabs/sui/issues/1837

**Pull request merged**
